src/utils_request.js

```diff
@@ -0,0 +1,354 @@
+ */
+export const requestExtended = async (options) => {
+    suppressTunnelAgentAssertError();
+
+    return new Promise((resolve, reject) => {
+        const opts = _.defaults({}, options, REQUEST_EXTENDED_DEFAULT_OPTIONS);
+
+        const {
+            abortFunction,
+            throwOnHttpError,
+        } = opts;
+        const method = opts.method.toLowerCase();
+
+        // Using the streaming API of Request to be able to
+        const request = rqst[method](opts);
+        request
+            .on('error', err => reject(err))
+            .on('response', async (res) => {
+                let shouldAbort;
+
+                try {
+                    shouldAbort = abortFunction(res);
+                } catch (e) {
+                    reject(e);
+                }
+
+                if (shouldAbort) {
+                    request.abort();
+                    res.destroy();
+
+                    return reject(new Error(`utils.requestBetter: Request for ${opts.url} aborted due to abortFunction`));
+                }
+
+                let cType;
+                try {
+                    cType = contentType.parse(res);
+                } catch (err) {
+                    res.destroy();
+                    // No reason to parse the body if the Content-Type header is invalid.
+                    return reject(new Error(`utils.requestBetter: Invalid Content-Type header for URL: ${request.url}`));
+                }
+
+                const { encoding } = cType;
+
+                // 5XX and 4XX codes are handled as errors, requests will be retried.
+                const status = res.statusCode;
+
+                if (status >= 400 && throwOnHttpError) {
+                    const error = await getMoreErrorInfo(res, cType);
+                    reject(error);
+                }
+
+                // Content-Type is fine. Read the body and respond.
+                try {
+                    res.body = await readStreamIntoString(res, encoding);
+                    resolve(res);
+                } catch (err) {
+                    // Error in reading the body.
+                    reject(err);
+                }
+            });
+    });
+};
+
+/**
+ * Gets more info about the error.
+ * Errors are often sent as JSON, so attempt to parse them, despite Accept header being set to something different.
+ * @param {http.IncomingMessage} response
+ * @param {Object} cType
+ * @param {String} cType.type
+ * @param {String} cType.encoding
+ * @returns {Promise<Error>}
+ */
+async function getMoreErrorInfo(response, cType) {
+    const { type, encoding } = cType;
+    const { status } = response;
+    let body;
+    try {
+        body = await readStreamIntoString(response, encoding);
+    } catch (err) {
+        // Error in reading the body.
+        return err;
+    }
+
+    if (type === 'application/json') {
+        const errorResponse = JSON.parse(body);
```

Comment:
IMHO we should have a try-catch here and if it throws then use the truncated message as error msg. Because it's quite common that in a case of network error the body might be incomplete and cannot be parsed. The error that gets thrown then will be confusing.